zboxcore/sdk/repairCallback.go

```diff
@@ -25,39 +25,36 @@ func (s *StatusBar) Completed(allocationId, filePath string, filename string, mi
 
 func (s *StatusBar) Error(allocationID string, filePath string, op int, err error) {
 	s.success = false
-	defer s.wg.Done()
-	defer mutUnlock(s.allocID)
-
-	var errDetail interface{} = "Unknown Error"
-	if err != nil {
-		errDetail = err.Error()
-	}
-
-	l.Logger.Error("Error in status callback. Error = ", errDetail)
+	s.err = err
+	l.Logger.Error("Error in status callback. Error = ", err.Error())
 }
 
 func (s *StatusBar) RepairCompleted(filesRepaired int) {
+	if s.err == nil {
+		s.success = true
+	}
 	defer s.wg.Done()
-	s.success = true
 	mutUnlock(s.allocID)
 	l.Logger.Info("Repair completed. Files repaired = ", filesRepaired)
+
 }
 
 type StatusBar struct {
 	wg      *sync.WaitGroup
 	allocID string
 	success bool
+	err     error
 }
 
 func NewRepairBar(allocID string) *StatusBar {
+	mapLock.Lock()
 	if _, ok := mutMap[allocID]; !ok {
-		mapLock.Lock()
 		mutMap[allocID] = &sync.Mutex{}
-		mapLock.Unlock()
 	}
 	if !mutMap[allocID].TryLock() {
 		return nil
 	}
+	mapLock.Unlock()
```

Comment:
This unlock won't be called if the code execution goes in line 55. Probably it's better to use `defer mutMapLock.Unlock()` after line 50?

Reply:
I think even now this might be a problem. Should we instead use ReadWriterLock for read and write separately?

Reply:
We can defer it and it should be fine, write will take no time so should be alright.

Reply:
We would also need to check if this is even a good thing to do and this does not cause starvation to other threads :( 
I think using read lock and then the write lock might NOT be a good idea.